File: apps/web/types/ideas.ts
export type NoveltyLevel = "aligned" | "stretch" | "experimental";
export type SimilarityCategory = "new" | "related_but_distinct" | "follow_up" | "too_similar";
export type IdeaStatus = "proposed" | "saved" | "archived";
export type ProductionEffort = "low" | "medium" | "high";
export type FeedbackRating = "love" | "maybe" | "not_for_me" | "already_covered";

export const FEEDBACK_RATINGS: { value: FeedbackRating; label: string }[] = [
  { value: "love", label: "Love it" },
  { value: "maybe", label: "Maybe" },
  { value: "not_for_me", label: "Not for me" },
  { value: "already_covered", label: "Already covered" },
];

export const SCRIPT_MODES: { value: string; label: string }[] = [
  { value: "polished_explainer", label: "Polished explainer" },
  { value: "casual_talking_head", label: "Casual talking head" },
  { value: "grwm_story", label: "GRWM story" },
  { value: "news_recap", label: "News recap" },
  { value: "comment_response", label: "Comment response" },
  { value: "personal_story", label: "Personal story" },
];

export interface ContentIdeaOut {
  id: string;
  title: string;
  concept: string;
  content_pillar: string;
  recommended_format: string;
  format_label_lt: string;
  why_it_fits_me: string;
  inspiration_pattern: string | null;
  originality_note: string | null;
  closest_existing_post_id: string | null;
  similarity_score: number;
  similarity_category: SimilarityCategory;
  novelty_level: NoveltyLevel;
  hook_options: string[];
  outline: string[];
  suggested_duration_seconds: number;
  production_effort: ProductionEffort;
  output_language: string;
  status: IdeaStatus;
  created_at: string;
}

export interface GeneratedBriefOut {
  id: string;
  idea_id: string;
  objective: string;
  target_viewer: string;
  promise: string;
  recommended_format: string;
  recommended_duration_seconds: number;
  hook_choices: string[];
  beats: { label: string; description: string }[];
  b_roll_suggestions: string[];
  on_screen_text: string[];
  editing_notes: string[];
  closing_line: string;
  call_to_action: string;
  caption_options: string[];
  hashtags: string[];
  claims_to_verify: string[];
  output_language: string;
}

export interface GeneratedScriptOut {
  id: string;
  idea_id: string;
  mode: string;
  spoken_lines: string[];
  editing_notes: string[];
  estimated_duration_seconds: number;
  placeholders: string[];
  output_language: string;
}

export interface IdeaGenerateRequest {
  count?: number;
  content_pillar?: string | null;
  recommended_format?: string | null;
  instructions?: string | null;
  excluded_subjects?: string[];
  output_language?: string | null;
}

export interface IdeaFeedbackRequest {
  rating: FeedbackRating;
  comment?: string | null;
}

export interface ScriptGenerateRequest {
  mode: string;
}

File: apps/api/app/schemas/ideas.py
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class ContentIdeaOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    title: str
    concept: str
    content_pillar: str
    recommended_format: str
    format_label_lt: str
    why_it_fits_me: str
    inspiration_pattern: str | None
    originality_note: str | None
    closest_existing_post_id: uuid.UUID | None
    similarity_score: float
    similarity_category: str
    novelty_level: str
    hook_options: list[str]
    outline: list[str]
    suggested_duration_seconds: int
    production_effort: str
    output_language: str
    status: str
    created_at: datetime


class GeneratedBriefOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    idea_id: uuid.UUID
    objective: str
    target_viewer: str
    promise: str
    recommended_format: str
    recommended_duration_seconds: int
    hook_choices: list[str]
    beats: list[dict]
    b_roll_suggestions: list[str]
    on_screen_text: list[str]
    editing_notes: list[str]
    closing_line: str
    call_to_action: str
    caption_options: list[str]
    hashtags: list[str]
    claims_to_verify: list[str]
    output_language: str


class GeneratedScriptOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    idea_id: uuid.UUID
    mode: str
    spoken_lines: list[str]
    editing_notes: list[str]
    estimated_duration_seconds: int
    placeholders: list[str]
    output_language: str


class ImageResultOut(BaseModel):
    url: str
    thumbnail_url: str
    source_url: str
    credit: str | None = None


class MemeResultOut(BaseModel):
    url: str
    template_name: str
    caption_lines: list[str]


class IdeaSourcedMediaOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    idea_id: uuid.UUID
    images: list[ImageResultOut]
    memes: list[MemeResultOut]


class IdeaGenerateRequest(BaseModel):
    count: int = Field(default=3, ge=1, le=10)
    content_pillar: str | None = None
    recommended_format: str | None = None
    instructions: str | None = None
    excluded_subjects: list[str] = Field(default_factory=list)
    output_language: str | None = None


class IdeaFeedbackRequest(BaseModel):
    rating: str
    comment: str | None = None


class ScriptGenerateRequest(BaseModel):
    mode: str
